[jwyk] iP

jwyk opened the pull request.

jwyk committed:
Change chatbot name to Bob

jwyk committed:
Rename chatbot, implement greeting and exit.

jwyk committed:
Implement Bob echoing commands and exit functionality

jwyk committed:
Added ability to store user texts as tasks and return a list when requested

jwyk committed:
Added ability to mark and unmark tasks, as well as implement Task Class

jwyk committed:
Refactor Task printing with getTask function and add Deadline, Event class skeletons

jwyk committed:
Add the ability to parse Deadlines, Events and Tasks.

LeThiHongMinh commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,113 @@
+
+    //Initializer for Bob
+    public Bob() {
+        current = 0;
```
The variable 'current' is used to track the number of tasks. It would be clearer if it were renamed to something more descriptive, like 'taskCount', to improve readability and understanding of its role.

LeThiHongMinh commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,113 @@
+            command = input;
+        }
+
+        switch (command) {
```
I like how you use "switch" statement for handling different commands, making it clear and easy to extend if new commands need to be added in the future.

LeThiHongMinh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
```
You've used methods like getStatusIcon() to retrieve the task status in a clean, reusable manner, which is a good practice to avoid redundant code.

LeThiHongMinh commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,113 @@
+                System.out.println(i + 1 + ". " + list[i].getTask());
+            }
+            break;
+        case "mark":
```
The "mark" and "unmark" may follow a similar pattern, which may need refactoring.

LeThiHongMinh commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,113 @@
+    //Initializer for Bob
+    public Bob() {
+        current = 0;
+        list = new Task[100];
```
It's good practice to replace "100" with a named constant to indicate its purpose such as "MAX_TASKS"

LeThiHongMinh reviewed (commented):
Great work overall! 👍

jwyk committed:
Refactor Bob to use arraylist instead of fixed size array and fixed minor printing inconsistencies between tasks

melee01 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,114 @@
+            System.out.println("Current number of tasks: " + current);
+            break;
+
+        case "todo":
```
Better to have todo as its own class like how you separate it in Deadline and Event, if there is any additional guideline on modifying ToDo, your current code will need to be changed a lot.

melee01 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,114 @@
+
+    //Initializer for Bob
+    public Bob() {
+        current = 0;
```
Current variable is indicated to tell on how many tassks are the in the list, however, it is best to have a meaningful definition to every variable you used, it will improve readability, especially when you reviewed and go through your own code, you can grasp and make changes easier.

melee01 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,114 @@
+import java.util.Scanner;
+import java.util.ArrayList;
+
```
I think you can add another layering in interface, such as home page. This will help in case when another additional feature that is outside of this Bob chatbot functionality, where it works alongside this Bob.java, the code will be easier for extension later on. So, can create something like home page, and then call Bob.java inside it. It is neater as well for me.

melee01 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,114 @@
+            command = input;
+        }
+
+        switch (command) {
```
Using Switch and differentiate inputs by cases is smart here, I appreciate it so much.

melee01 reviewed (commented):
Overall, it is written well although some improvements can be made.

jwyk committed:
Refactor getTask function to a generic toString class method to reduce verbosity in code

jwyk committed:
Refactor code to replace current with Arraylist.size()

jwyk committed:
Added support for input testing and squashed minor output bugs

jwyk committed:
Added support for input testing and squashed minor output bugs

jwyk committed:
Add Javadocs for Deadline, Event and Task classes

jwyk committed:
Refactor Code to do the following
Add Parser class for handling string inputs in a proper class.
Modified Task into an abstract class, and added Todo class that properly abstracts Task.

jwyk committed:
Change cases for switch from a string to a more readable constant

jwyk committed:
Create a BobException class for custom errors

jwyk committed:
Add support for BobException Errors

jwyk committed:
Refactor Bob to use getInput for error handling and rewrite mark and unmark error handling

jwyk committed:
Refactor Deadline, Todo and Event to handle proper errors

jwyk committed:
Fix logical error check for Events, and added tests for error checking

jwyk committed:
Merge branch 'branch-Level-5'

jwyk committed:
Change compilation line to account for packages

jwyk committed:
Rearrange classes into java packages

jwyk committed:
Merge branch 'branch-A-Packages'

jwyk committed:
Add the ability to delete tasks

jwyk committed:
Add the ability to save tasks using Storage class

jwyk committed:
Add get functions for description, due date, start date and end date

jwyk committed:
Add the ability to load tasks

jwyk committed:
Merge branch 'branch-Level-6'

jwyk committed:
Merge branch 'branch-Level-7'

jwyk committed:
Change file save location

jwyk committed:
Fix logic error where delete operations were not saving

Paulifyer commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,202 @@
+            String[] eventArray = input.substring(taskPosition).split("/from ", 2); //Split the description and event
+            if (eventArray.length != 2) {
+                throw new BobException("Okay we are checking... there's no from." +
+                        "Please add /from to demarcate the duration of the event.");
+            }
+
+            String[] fromToArray = eventArray[1].split(" /to", 2);
+            //This error check needs to be split with the /from check, or you will get index out of bounds
+            if (fromToArray.length != 2) {
+                throw new BobException("Okay we are checking... there's no to." +
+                        "Please add /to to demarcate the duration of the event.");
+            }
+
+            String eventName = eventArray[0].trim();
+            String startDate = fromToArray[0].trim();
+            String endDate = fromToArray[1].trim();
+            taskList.add(new Event(eventName, startDate, endDate));
+            System.out.println("added: " + taskList.get(taskList.size() - 1));
+            System.out.println("Current number of tasks: " + taskList.size());
+            Storage.save(taskList);
+            break;
+
+        case COMMAND_TODO:
+            if (taskPosition == -1) {
+                throw new BobException("Okay we are checking... there's no description." +
+                        "Please add a description."); //Missing description
+            }
+            input = input.substring(taskPosition + 1);
+            taskList.add(new Todo(input));
+            System.out.println("added: " + taskList.get(taskList.size() - 1));
+            System.out.println("Current number of tasks: " + taskList.size());
+            Storage.save(taskList);
+            break;
+
+        default:
+            //None of the above commands met
+            System.out.println("Sorry, I didn't get that. Will come back to you...");
+        }
+        System.out.println("____________________________________________________________");
+    }
```
I think you can shorten this method by breaking up each case into individual methods?

Paulifyer commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * String constants for each command type
+     */
+    private static final String COMMAND_BYE = "bye";
+    private static final String COMMAND_LIST = "list";
+    private static final String COMMAND_MARK = "mark";
+    private static final String COMMAND_UNMARK = "unmark";
+    private static final String COMMAND_DEADLINE = "deadline";
+    private static final String COMMAND_TODO = "todo";
+    private static final String COMMAND_EVENT = "event";
+    private static final String COMMAND_DELETE = "delete";
```
Instead of using strings here you could maybe use an enumerator?

Paulifyer commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     * @param list ArrayList containing Tasks
+     */
+
+    public static void save(ArrayList<Task> list) throws BobException, IOException {
```
Technically you do not need to throw IOException here as your code already catches it

Paulifyer commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     * Returns whether a save file exists or not
+     */
+
+    public static boolean exists() {
```
The naming of this method could be more clear

jwyk committed:
Refactor Bob to utilize new TaskList class

jwyk committed:
Update testing files to auto delete bob.txt in order to prevent replication bug

jwyk committed:
Refactored the code to use More OOP

- Ui: Prints out all UI messages related to commands.
- Command: Handle each Command action.
- Action: Map each input type to a command type.
- Parser: Deals with user inputs and mappings.
- TaskList: Rewrite to contain the task list and related operations.
- Storage: Handles the saving and loading of tasklist data.

jwyk committed:
Add ability to parse dates and times to events and deadlines

jwyk committed:
Remove unused parameters and fix start date used for end dates in events

jwyk committed:
Merge pull request #1 from jwyk/branch-Level-8

Add ability to store date and time

jwyk committed:
Merge pull request #2 from jwyk/master

Add ability to find tasks

jwyk committed:
Revert "Add ability to find tasks"

jwyk committed:
Merge pull request #3 from jwyk/revert-2-master

Revert "Add ability to find tasks"

jwyk committed:
Add the ability to find tasks based on their description

jwyk committed:
Merge pull request #4 from jwyk/branch-Level-9

Add ability to find tasks

jwyk committed:
Revert "Add ability to find tasks"

jwyk committed:
Merge pull request #5 from jwyk/revert-4-branch-Level-9

Revert "Add ability to find tasks"

jwyk committed:
Add the ability to find tasks based on their description

jwyk committed:
Merge pull request #6 from jwyk/branch-Level-9

Add the ability to find tasks

jwyk committed:
Refactor codebase to incorporate more packages

jwyk committed:
Add Javadocs

- Introduce Javadocs to each method and class
- Rewrite existing comments to fit proper standards

jwyk committed:
Merge pull request #7 from jwyk/branch-A-JavaDoc

jwyk committed:
Add User Guide and update README to be Bob instead of Duke
